src/query/storage/m3/cluster_resolver.go

```diff
@@ -80,6 +80,45 @@ func resolveClusterNamespacesForQuery(
 	clusters Clusters,
 	opts *storage.FanoutOptions,
 	restrict *storage.RestrictQueryOptions,
+) (consolidators.QueryFanoutType, ClusterNamespaces, error) {
+	// 1. First resolve the logical plan.
+	fanout, namespaces, err := resolveClusterNamespacesForQueryLogicalPlan(now,
+		start, end, clusters, opts, restrict)
+	if err != nil {
+		return fanout, namespaces, err
+	}
+
+	// 2. Create physical plan.
+	// Now de-duplicate any namespaces that might be fetched twice due to
+	// the fact some of the same namespaces are reused once for unaggregated
+	// and another for aggregated rollups (which don't collid with timeseries).
+	filtered := namespaces[:0]
+	for _, ns := range namespaces {
+		keep := true
+		// Small enough that we can do n^2 here instead of creating a map,
+		// usually less than 4 namespaces resolved.
+		for _, existing := range filtered {
+			if ns.NamespaceID().Equal(existing.NamespaceID()) {
```

Comment:
This should be sufficient for equality right? Don't think we can have two namespaces with the same ID, just double checking

Reply:
Actually, this may need some sort of logic where if IDs match, preferentially take unagg -> smaller retention -> larger retention, as the namespace to look up for that ID, in that order

Reply:
Yeah the namespace registry is a map keyed by namespace ID so this should be good here 👍